### PR TITLE
Add `makeEthereumJSONRPCRequest` to WalletLinkRelay

### PR DESCRIPTION
**Pull request by cb-jake**
## Description

* Add flexibility to handleAsyncMethods. Move making json rpc request into walletlink relay

### CODE REVIEW

**Review by uri44derr13 — COMMENTED** [review]
#

> 

>  @`
Screenshot_20211217-200604_Cryptocom
Duplicate of #`

File: js/src/relay/WalletLinkRelay.ts [hunk 1]
```diff
@@ -840,4 +848,29 @@ export class WalletLinkRelay implements WalletLinkRelayAbstract {
       }
     })
   }
+
+  public async makeEthereumJSONRPCRequest(
+    request: JSONRPCRequest,
+    jsonRpcUrl: string
```
Comment (hieronymus777) [inline js/src/relay/WalletLinkRelay.ts:854]: This still requires that the jsonRpcUrl is set.  I think its preferable to make the jsonRpcUrl optional.  Let's make this call fail if it's not provided.

Comment (cb-jake) [reply js/src/relay/WalletLinkRelay.ts:854]: Good callout. I left throwing the error inside the provider. It makes more sense to have it here.

Comment (uri44derr13) [reply js/src/relay/WalletLinkRelay.ts:854]: (3301:5214:205ce2997062)

File: js/src/provider/WalletLinkProvider.ts [hunk 2]
```diff
@@ -638,28 +653,8 @@ export class WalletLinkProvider
         return this._wallet_switchEthereumChain(params)
     }
 
-    if (!this.jsonRpcUrl) throw Error("Error: No jsonRpcUrl provided")
```
Comment (cb-jake) [inline js/src/provider/WalletLinkProvider.ts:641]: Moved this into the WalletLinkRelay

File: js/src/provider/WalletLinkProvider.ts [hunk 3]
```diff
@@ -4,23 +4,34 @@
 
 import SafeEventEmitter from "@metamask/safe-event-emitter"
 import BN from "bn.js"
-import { ethErrors, serializeError } from "eth-rpc-errors"
+import { ethErrors } from "eth-rpc-errors"
 
 import { WalletLinkAnalytics } from "../connection/WalletLinkAnalytics"
-import { EthereumChain } from '../EthereumChain'
+import { EthereumChain } from "../EthereumChain"
```
Comment (cb-jake) [inline js/src/provider/WalletLinkProvider.ts:10]: Theres a lot of prettier/lint changes that didn't make it in last time. I commented where the actual changes in this file are.

File: js/src/provider/WalletLinkProvider.ts [hunk 4]
```diff
@@ -638,28 +653,8 @@ export class WalletLinkProvider
...
-        const { error } = response
-
-        if (error) {
-          throw serializeError(error)
-        }
-
-        return response
-      })
+    const relay = await this.initializeRelay()
+    return relay.makeEthereumJSONRPCRequest(request, this.jsonRpcUrl)
```
Comment (cb-jake) [inline js/src/provider/WalletLinkProvider.ts:657]: Added to the walletlink relay abstract

**Comment by cb-jake** [comment]
@hieronymus777 made the suggested updates. Looks like prettier caught some of the files and added some noise.

File: js/src/provider/WalletLinkProvider.ts (hunk 3 above)
Comment (uri44derr13) [reply js/src/provider/WalletLinkProvider.ts:10]: Uri44derr

**Comment by ryan22271** [comment]
You crooks are reported and responcible


Sent from ME


On Monday, December 20, 2021, 1:44 PM, Alex ***@***.***> wrote:




@uri44derr13 commented on this pull request.

In js/src/provider/WalletLinkProvider.ts:
>  
 import { WalletLinkAnalytics } from "../connection/WalletLinkAnalytics"
-import { EthereumChain } from '../EthereumChain'
+import { EthereumChain } from "../EthereumChain"

Uri44derr

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

File: js/src/provider/WalletLinkProvider.ts (hunk 3 above)
Comment (uri44derr13) [reply js/src/provider/WalletLinkProvider.ts:10]: Thank you
fb bin (1)

**Review by uri44derr13 — COMMENTED** [review]
#``

File: js/src/relay/WalletLinkRelay.ts (hunk 1 above)
Comment (hieronymus777) [reply js/src/relay/WalletLinkRelay.ts:854]: Sorry I mean can we make jsonRpcUrl optional.

Like
```
public async makeEthereumJSONRPCRequest(
    request: JSONRPCRequest,
    jsonRpcUrl?: string
)
```

Comment (hieronymus777) [reply js/src/relay/WalletLinkRelay.ts:854]: Also can we add a return type for this?

Comment (cb-jake) [reply js/src/relay/WalletLinkRelay.ts]: I just pushed the change for that!

Comment (cb-jake) [reply js/src/relay/WalletLinkRelay.ts]: of course! let me add the return type

**Review by asilkans — APPROVED** [review]

**Comment by uri44derr** [comment]
> ## Description
> * Add flexibility to handleAsyncMethods. Move making json rpc request into walletlink relay

**Comment by uri44derr** [comment]
Sooo

File: js/src/relay/WalletLinkRelayAbstract.ts [hunk 5]
```diff
@@ -74,20 +78,51 @@ export abstract class WalletLinkRelayAbstract {
...
 
   abstract setAppInfo(appName: string, appLogoUrl: string | null): void
+
   abstract setAccountsCallback(
-    accountsCallback: (accounts: [string]) => void
+    accountsCallback: (accounts: string[]) => void
   ): void
+
   abstract setChainCallback(
     chainIdCallback: (chainId: string, jsonRpcUrl: string) => void
```
Comment (hieronymus777) [inline js/src/relay/WalletLinkRelayAbstract.ts:98]: Let's also please make the `jsonRpcUrl` optional in this callback.

**Review by hieronymus777 — APPROVED** [review]

**Pull request merged**